src/handlers/publish.rs

```diff
@@ -0,0 +1,124 @@
+    db: &Database,
+    request: &PublishRequest,
+    token: &ApiToken,
+) -> Result<PublishInfo, PublishError> {
+    // Parse the forc manifest file and verify that the version is set.
+    let upload = db.conn().get_upload(request.upload_id)?;
+    // For now, only package manifests are supported. Workspace manifests will be supported in the future.
+    let forc_manifest = upload
+        .forc_manifest
+        .ok_or(PublishError::MissingForcManifest)?;
+    let pkg_manifest = PackageManifest::from_string(forc_manifest)
+        .map_err(|e| PublishError::InvalidForcManifest(e.to_string()))?;
+    let pkg_version = pkg_manifest
+        .project
+        .version
+        .ok_or(PublishError::InvalidForcManifest(
+            "Project manifest must have a version".to_string(),
+        ))?;
+
+    // Validate the package dependencies.
+    let package_deps = match pkg_manifest.dependencies {
+        Some(deps) => {
+            let mut package_deps = vec![];
+            for (name, dependency) in deps {
+                // Check if the package version exists in the database.
+                let version = dependency
+                    .version()
+                    .ok_or(PublishError::InvalidForcManifest(
+                        "Dependency must have a version".to_string(),
+                    ))?;
+                let _ = db
+                    .conn()
+                    .get_package_version(name.clone(), version.to_string())?;
+
+                package_deps.push(PartialPackageDep {
+                    dependency_package_name: name.clone(),
+                    dependency_version_req: version.to_string(),
+                });
+            }
+            package_deps
+        }
+        None => vec![],
+    };
+
+    // Insert package version into the database along with metadata from the package manifest.
+    let publish_info = PublishInfo {
+        package_name: pkg_manifest.project.name.clone(),
+        upload_id: request.upload_id,
+        num: pkg_version,
+        package_description: pkg_manifest.project.description.clone(),
+        repository: pkg_manifest.project.repository.clone(),
+        documentation: pkg_manifest.project.documentation.clone(),
+        homepage: pkg_manifest.project.homepage.clone(),
+        urls: request.urls.clone().unwrap_or_default(),
+        readme: upload.readme.clone(),
+        license: Some(pkg_manifest.project.license.clone()),
+    };
+    let package_version = db.conn().new_package_version(token, &publish_info)?;
+
+    // Insert package dependencies into the database.
+    let new_package_deps = package_deps
+        .iter()
+        .map(|dep| NewPackageDep {
+            dependent_package_version_id: package_version.id,
+            dependency_package_name: dep.dependency_package_name.clone(),
+            dependency_version_req: dep.dependency_version_req.clone(),
+        })
+        .collect();
+    let _ = db.conn().insert_dependencies(new_package_deps)?;
```

Comment:
nit: logging after successful publish
```suggestion
    info!("Successfully published package {} version {}", publish_info.package_name, publish_info.num);

    let _ = db.conn().insert_dependencies(new_package_deps)?;
```